vault/proxy.go

```diff
@@ -0,0 +1,296 @@
+package vault
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	socks5 "github.com/armon/go-socks5"
+	isatty "github.com/mattn/go-isatty"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/knownhosts"
+	"golang.org/x/net/http/httpproxy"
+)
+
+type ProxyRouter struct {
+	proxyConf httpproxy.Config
+}
+
+func (n ProxyRouter) Proxy(req *http.Request) (*url.URL, error) {
+	return n.proxyConf.ProxyFunc()(req.URL)
+}
+
+func NewProxyRouter() (*ProxyRouter, error) {
+	httpProxy := getEnvironmentVariable("HTTP_PROXY", "http_proxy")
+	httpsProxy := getEnvironmentVariable("HTTPS_PROXY", "https_proxy")
+
+	allProxy := getEnvironmentVariable("SAFE_ALL_PROXY", "safe_all_proxy")
+	if allProxy != "" {
+		httpProxy = allProxy
+		httpsProxy = allProxy
+	}
+
+	noProxy := getEnvironmentVariable("NO_PROXY", "no_proxy")
+
+	knownHostsFile := getEnvironmentVariable("KNOWN_HOSTS_FILE", "known_hosts_file")
+	skipHostKeyString := getEnvironmentVariable("SKIP_HOST_KEY_VALIDATION", "skip_host_key_validation")
+	skipHostKeyValidation := skipHostKeyString != "" && skipHostKeyString != "false"
+
+	var tunnelsToOpen []*string
+	if strings.HasPrefix(httpProxy, "ssh+socks5://") {
+		tunnelsToOpen = append(tunnelsToOpen, &httpProxy)
+	}
+
+	if strings.HasPrefix(httpsProxy, "ssh+socks5://") {
+		tunnelsToOpen = append(tunnelsToOpen, &httpsProxy)
+	}
+
+	for i := range tunnelsToOpen {
+		//If we haven't already opened a proxy for this...
+		if i == 0 || *(tunnelsToOpen[i]) != *(tunnelsToOpen[i-1]) {
+			//Let's open a proxy!
+			u, err := url.Parse(*(tunnelsToOpen[i]))
+			if err != nil {
+				return nil, fmt.Errorf("Could not parse proxy URL (%s): %s", *tunnelsToOpen[i], err)
+			}
+
+			if u.User == nil {
+				return nil, fmt.Errorf("No user provided for SSH proxy")
+			}
+
+			sshClient, err := StartSSHTunnel(SOCKS5SSHConfig{
+				Host:                  u.Host,
+				User:                  u.User.Username(),
+				PrivateKey:            u.Path,
+				KnownHostsFile:        knownHostsFile,
+				SkipHostKeyValidation: skipHostKeyValidation,
+			})
+			if err != nil {
+				return nil, fmt.Errorf("Could not start SSH tunnel: %s", err)
+			}
+
+			socks5Addr, err := StartSOCKS5Server(sshClient.Dial)
+			if err != nil {
+				return nil, fmt.Errorf("Could not start SOCKS5 Server: %s", err)
+			}
+			*tunnelsToOpen[i] = fmt.Sprintf("socks5://%s", socks5Addr)
+		} else {
+			//Let's get the address of the already opened proxy
+			*tunnelsToOpen[i] = *tunnelsToOpen[i-1]
+		}
+	}
+
+	return &ProxyRouter{
+		proxyConf: httpproxy.Config{
+			HTTPProxy:  httpProxy,
+			HTTPSProxy: httpsProxy,
+			NoProxy:    noProxy,
+		},
+	}, nil
+}
+
+func getEnvironmentVariable(variables ...string) string {
+	var ret string
+	for _, v := range variables {
+		ret := os.Getenv(v)
+		if ret != "" {
+			break
```

Comment:
Can't you just `return ret` here, and then `return ""` at the end of the function?  Also, FWIW, os.LookupEnv returns `(string, bool)` so you can determine if the value is set; not sure if we have covered all the edge cases where users explicitly set the env var to "", and expected different behavior than when they didn't set it in the first place.

Reply:
:+1: